akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -665,7 +670,20 @@ object Flow {
    * '''Cancels when''' downstream cancels
```

Comment:
@seglo , notice this is one aspect of the operator's behaviour that has slightly changed.
an 'early' cancellation is stashed until the inner flow is materialized, it is then handed over to the materialized flow which can decide on handling the signal.
as mentioned in an earlier comment of mine, this behaviour is inspired by `flatMapPrefix` (and review comments+directions by @jrudolph ), I think we should also provide some fine grained control over this behaviour by specifying attributes. 

given that most flows actually cancels when downstream does, this change of behaviour is usually unnoticeable as long as upstream emits/completes and the future actually completes in a reasonable time.

Reply:
Thanks for pointing this out.  I think it would make sense to preserve this behaviour for `lazyFutureFlow` to avoid more possible regressions for existing users.  

Are you suggesting creating a new attribute to control whether operators that use `FutureFlow` should eagerly cancel the stage `onDownstreamFinish(Throwable)` (default case), or pass the signal through the inner stream materialization?  There's already a `CancellationStrategy` attribute, perhaps we could add a strategy to it.

Reply:
I'm basically talking about an attribute that controls this behaviour, it's needed both for `FutureFlow` and `FlatMapPrefix` as both currently 'intercepts' the downstream completion and delays its handling until there's a nested flow to delegate to. I also think the current behaviour should **not** be the default.
re. the existing `CancellationStrategy` attribute, it seems to be related to the way stages are cancelled so I'm not sure it makes sense to reuse it.
I'm also not sure if the same attribute can be used to control both stages, since it may make it difficult to specify for flows that combine both, on the other hand it seems that these kind of combinations (i.e. the lazyXXX family) would usually apply the same policy to both.

do u think we should implement this as part of this PR or in a separate one? I guess it boils down to the importance of maintaining the old behaviour.
please advice.

Reply:
I tend to say that relying on the previous behavior is somewhat brittle and already required careful design to be free of race conditions. To observe the new behavior you must have control over 1. when the future is completed and 2. the cancellation is received. In most cases, both of these are asynchronous and one can happen before the other.

The most likely case that I can think of is that you know 1. that cancellation was triggered and will arrive eventually and 2. the future flow is never completed. In particular, there's the case that the future is left hanging uncompleted accidentally. In this case the previous behavior would have prevented leaking the upstream. That sounds significant but even that is somewhat expected as never completing futures is known to lead to leaks.

Any other case how you would observe it?

Reply:
I think the current design eliminates the race condition, at least as a corruption. the stage is notified by 'wiring' Future.onComplete to an asyncCallback and internal state is updated only in the stage's context 
I guess u can call it a race condition since the stage may become non-deterministic, is this what u mean @jrudolph ? if so do u prefer the current implementation as proposed in this PR?

I think it's also worth noting that most flows simply propagate the downstream cancellation upstream, which results with the in port being closed. one downside of my current implementation is that in this common case the downstream/cancellation/failure signal is 'delayed'. since I believe this is the common case I think it should also be the default behaviour or at least possibly enabled using an attribute.

Reply:
> I think it's also worth noting that most flows simply propagate the downstream cancellation upstream, which results with the in port being closed. one downside of my current implementation is that in this common case the downstream/cancellation/failure signal is 'delayed'.

That's a good point. Another argument for the old behavior by default would be that you usually don't need the extra flexibility of having the flow react specially on cancellation. In almost all cases, the flow will just propagate the cancellation.

For `flatMapPrefix` it might be a bit different because there you'd might want to execute a side-effect when the prefix is collected regardless of whether the downstream was already cancelled or not.

I talked with @seglo about that whole issue and as much as we'd like to avoid the complexity of introducing an extra attribute/parameter to configure this, it seems it actually is needed.

Because most usages will not depend on holding back cancellation we'd say that the old behavior would be the default (for both `futureFlow` and `flatMapPrefix` for consistency) and if you want the extra flexibility of having the flow handle cancellation you would have to add an attribute as you suggested. It seems the extra complexity in the the stage implementations would be low, mostly a small change where instead of stashing the downstream cause you would eagerly cancel the stage.

WDYT, @eyalfa?

So, the changes needed are:

 * add the new attribute
 * implement the feature in FutureFlow and FlatMapPrefix
 * add/adapt tests to test both behaviors
 * document the attribute in scaladoc
 * (maybe later) make sure the stream operation documentation pages are adapted accordingly (we are thinking about having an extra page to document the features and differences between all those flow construction helpers)

Reply:
@jrudolph ,
this basically makes sense, few questions come to mind (on a first read, there may be more later on):
1. same attribute for both stages?
2. default should be the 'old behaviour' hence, if downstream cancels before future/prefix completion the stage cancels, otherwise the inner flow takes charge. right?
3. another possibility is 'going all the way' by always closing upstream upon downstream cancellation, even if the nested flow is already running. though in this case the nested flow may keep running although the outer stage is already closed... so I'm not sure if this is a good direction or not...

Reply:
I'll attempt to answer these, but we should wait for confirmation from @jrudolph before proceeding.

> same attribute for both stages?

Yes

> default should be the 'old behaviour' hence, if downstream cancels before future/prefix completion the stage cancels, otherwise the inner flow takes charge. right?

Good question. Johannes and I didn't discuss this use case specifically, but in an effort to preserve the old behaviour I think we should do this.

> another possibility is 'going all the way' by always closing upstream upon downstream cancellation, even if the nested flow is already running. though in this case the nested flow may keep running although the outer stage is already closed... so I'm not sure if this is a good direction or not...

I was thinking about this as well. I think it makes sense to support eagerly cancelling the outer stage regardless of the running state of the inner flow either as a replacement to the new behaviour, or in addition to. I think arguments could be made for all use cases. To support this behaviour in addition to the others we could add a strategy argument to the attribute to pick a behaviour.  This would give us 3 possible behaviours (or "strategies"):

1. Cancel outer stage if inner flow is not materialized yet (old behaviour)
2. Always propagate cancellation signal through inner flow
3. Cancel outer stage (and inner flow) despite run state of inner flow

Reply:
> I'll attempt to answer these, but we should wait for confirmation from @jrudolph before proceeding.
> 
> > same attribute for both stages?
> 
> Yes
> 
> > default should be the 'old behaviour' hence, if downstream cancels before future/prefix completion the stage cancels, otherwise the inner flow takes charge. right?
> 
> Good question. Johannes and I didn't discuss this use case specifically, but in an effort to preserve the old behaviour I think we should do this.
> 
> > another possibility is 'going all the way' by always closing upstream upon downstream cancellation, even if the nested flow is already running. though in this case the nested flow may keep running although the outer stage is already closed... so I'm not sure if this is a good direction or not...
> 
> I was thinking about this as well. I think it makes sense to support eagerly cancelling the outer stage regardless of the running state of the inner flow either as a replacement to the new behaviour, or in addition to. I think arguments could be made for all use cases. To support this behaviour in addition to the others we could add a strategy argument to the attribute to pick a behaviour. This would give us 3 possible behaviours (or "strategies"):
> 
> 1. Cancel outer stage if inner flow is not materialized yet (old behaviour)
> 2. Always propagate cancellation signal through inner flow
> 3. Cancel outer stage (and inner flow) despite run state of inner flow
3rd option complicates things, what about the materialized value? it has to be somehow propagated, tut the outer stage will be closed by that time... supporting this may result with a really awkward signature and contract.
I think it'd also be very difficult to explain and document, in addition most users will use option 1, very few will use 2nd option, not sure who's going to use the 3rd option.

Reply:
> I think it'd also be very difficult to explain and document, in addition most users will use option 1, very few will use 2nd option, not sure who's going to use the 3rd option.

That's true.  I thought that's what you meant by this comment, but maybe I misunderstood:

> another possibility is 'going all the way' by always closing upstream upon downstream cancellation, even if the nested flow is already running. though in this case the nested flow may keep running although the outer stage is already closed... so I'm not sure if this is a good direction or not...